File: assemble-packages.py
import os
import json

def byteify(input):
    if isinstance(input, dict):
        return {byteify(key): byteify(value)
                for key, value in input.iteritems()}
    elif isinstance(input, list):
        return [byteify(element) for element in input]
    elif isinstance(input, unicode):
        return input.encode('utf-8')
    else:
        return input
        
dirs = [
    'portal',
    'portal-api',
    'portal-chatbot',
    'portal-env',
    'portal-kong-adapter',
    'portal-mailer',
    'portal-kickstarter'    
]

base_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))

allDependencies = {}

for dirName in dirs:
    dir = os.path.join(base_dir, dirName)
    
    print 'Checking packages in: ' + dirName
    
    with open(os.path.join(dir, 'package.json')) as json_package:
        package = byteify(json.loads(json_package.read()))
    
    #print json.dumps(package)
    if 'dependencies' in package:
        for depName in package['dependencies']:
            depVersion = package['dependencies'][depName]
            if not depName in allDependencies:
                allDependencies[depName] = depVersion
            elif depVersion != allDependencies[depName]:
                print 'WARNING: dependency version mismatch for "' + dirName + '": ' + depName + ' - ' + depVersion + ' vs. ' + allDependencies[depName]
                if depVersion > allDependencies[depName]:
                    print 'WARNING: Taking newer version: ' + depVersion
                    allDependencies[depName] = depVersion

with open(os.path.join(os.path.dirname(__file__), 'package.json')) as package_json:
    this_package = byteify(json.loads(package_json.read()))
this_package['dependencies'] = allDependencies
print 'Writing all dependencies to package.all.json.'
with open(os.path.join(os.path.dirname(__file__), 'package.all.json'), 'w') as all_package_json:
    json.dump(this_package, all_package_json, indent=2, sort_keys=True)

File: assemble-packages.js
'use strict';

var fs = require('fs');
var path = require('path');

var dirs = [
    'wicked.portal',
    'wicked.portal-api',
    'wicked.portal-chatbot',
    'wicked.portal-env',
    'wicked.portal-kong-adapter',
    'wicked.portal-mailer',
    'wicked.portal-kickstarter',
    'wicked.portal-test/portal-api',
    'wicked.portal-test/portal-kong-adapter'
];

var baseDir = path.resolve(path.join(__dirname, '..'));

var allDependencies = {};

for (var i=0; i<dirs.length; ++i) {
    var dirName = dirs[i];
    var dir = path.join(baseDir, dirName);
    console.log('Checking packages.json in: ' + dir);
    
    var pkg = JSON.parse(fs.readFileSync(path.join(dir, 'package.json')));

    for(var depName in pkg.dependencies) {
        var depVersion = pkg.dependencies[depName];
        if (depVersion.startsWith('..'))
            continue;
        if (!allDependencies[depName])
            allDependencies[depName] = depVersion;
        else if (allDependencies[depName] != depVersion) {
            console.log('WARNING: Dependency version mismatch for "' + dirName + '": ' + depName + ' - ' + depVersion + ' vs. ' + allDependencies[depName]);
            if (depVersion > allDependencies[depName]) {
                console.log('WARNING: Taking newer version: ' + depVersion);
                allDependencies[depName] = depVersion;
            }
        }
    }    
}

// Re-add the portal-env we filtered out above
allDependencies['portal-env'] = '../portal-env.tgz';

var thisPackage = JSON.parse(fs.readFileSync(path.join(__dirname, 'package.json'), 'utf8'));
thisPackage.dependencies = allDependencies;

console.log(JSON.stringify(thisPackage, null, 2));

var allPackageFileName = path.join(__dirname, 'package.all.json');
console.log('Writing to ' + allPackageFileName); 
fs.writeFileSync(allPackageFileName, JSON.stringify(thisPackage, null, 2), 'utf8');
